internal/translation/customroles/conversion.go

```diff
@@ -0,0 +1,127 @@
+package customroles
+
+import (
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"go.mongodb.org/atlas-sdk/v20231115008/admin"
+)
+
+type CustomRole struct {
+	*akov2.CustomRole
+}
+
+func NewCustomRole(role *akov2.CustomRole) *CustomRole {
+	return &CustomRole{
+		CustomRole: role,
+	}
+}
+
+func toAtlas(role *CustomRole) *admin.UserCustomDBRole {
+	atlas := admin.NewUserCustomDBRoleWithDefaults()
+
+	atlas.SetRoleName(role.Name)
+
+	actions := make([]admin.DatabasePrivilegeAction, 0, len(role.Actions))
+	for _, action := range role.Actions {
+		resources := make([]admin.DatabasePermittedNamespaceResource, 0, len(action.Resources))
+		for _, resource := range action.Resources {
+			if resource.Cluster != nil && !*resource.Cluster {
+				resource.Cluster = nil
+			}
+			resources = append(resources, admin.DatabasePermittedNamespaceResource{
+				Collection: admin.GetOrDefault(resource.Collection, ""),
+				Db:         admin.GetOrDefault(resource.Database, ""),
+				Cluster:    admin.GetOrDefault(resource.Cluster, false),
+			})
+		}
+		actions = append(actions, admin.DatabasePrivilegeAction{
+			Action:    action.Name,
+			Resources: &resources,
+		})
+	}
+	atlas.SetActions(actions)
```

Comment:
Nitpick, I would move this to a sub-function, but maybe that is just me:

```suggestion
	atlas.SetActions(toAtlasActions(role.Actions))
```

and separately:
```go
func toAtlasActions(actions []akov2.Action) *admin.UserCustomDBRole {
actions := make([]admin.DatabasePrivilegeAction, 0, len(role.Actions))
	for _, action := range role.Actions {
		resources := make([]admin.DatabasePermittedNamespaceResource, 0, len(action.Resources))
		for _, resource := range action.Resources {
			if resource.Cluster != nil && !*resource.Cluster {
				resource.Cluster = nil
			}
			resources = append(resources, admin.DatabasePermittedNamespaceResource{
				Collection: admin.GetOrDefault(resource.Collection, ""),
				Db:         admin.GetOrDefault(resource.Database, ""),
				Cluster:    admin.GetOrDefault(resource.Cluster, false),
			})
		}
		actions = append(actions, admin.DatabasePrivilegeAction{
			Action:    action.Name,
			Resources: &resources,
		})
	}
	return actions
}
```

But again, that is a nitpick and maybe personal preference.

Reply:
I've separated these out into their own functions.